packages/protocol/contracts/team/TimeLockTokenPool.sol

```diff
@@ -0,0 +1,269 @@
+/// @title TimeLockTokenPool
+/// Contract for managing Taiko tokens allocated to different roles and
+/// individuals.
+///
+/// Manages Taiko tokens through a three-state lifecycle: "allocated" to
+/// "granted, owned, and locked," and finally to "granted, owned, and unlocked."
+/// Allocation doesn't transfer ownership unless specified by grant settings.
+/// Conditional allocated tokens can be canceled by invoking `void()`, making
+/// them available for other uses. Once granted and owned, tokens are
+/// irreversible and their unlock schedules are immutable.
+///
+/// We should deploy multiple instances of this contract for different roles:
+/// - investors
+/// - team members, advisors, etc.
+/// - grant program grantees
+
+contract TimeLockTokenPool is OwnableUpgradeable {
+    using SafeERC20Upgradeable for ERC20Upgradeable;
+
+    struct Grant {
+        uint128 amount;
+        // If non-zero, indicates the start time for the recipient to receive
+        // tokens, subject to an unlocking schedule.
+        uint64 grantStart;
+        // If non-zero, indicates the time after which the token to be received
+        // will be actually non-zero
+        uint64 grantCliff;
+        // If non-zero, specifies the total seconds required for the recipient
+        // to fully own all granted tokens.
+        uint32 grantPeriod;
+        // If non-zero, indicates the start time for the recipient to unlock
+        // tokens.
+        uint64 unlockStart;
+        // If non-zero, indicates the time after which the unlock will be
+        // actually non-zero
+        uint64 unlockCliff;
+        // If non-zero, specifies the total seconds required for the recipient
+        // to fully unlock all owned tokens.
+        uint32 unlockPeriod;
+    }
+
+    struct Recipient {
+        uint128 amountWithdrawn;
+        Grant[] grants;
+    }
+
+    uint256 public constant MAX_GRANTS_PER_ADDRESS = 8;
+
+    address public taikoToken;
+    address public sharedVault;
+    uint128 public totalAmountGranted;
+    uint128 public totalAmountVoided;
+    uint128 public totalAmountWithdrawn;
+    mapping(address recipient => Recipient) public recipients;
+    uint128[47] private __gap;
+
+    event Granted(address indexed recipient, Grant grant);
+    event Voided(address indexed recipient, uint128 amount);
+    event Withdrawn(address indexed recipient, address to, uint128 amount);
+
+    error INVALID_GRANT();
+    error INVALID_PARAM();
+    error NOTHING_TO_VOID();
+    error NOTHING_TO_WITHDRAW();
+    error TOO_MANY();
+
+    function init(
+        address _taikoToken,
+        address _sharedVault
+    )
+        external
+        initializer
+    {
+        OwnableUpgradeable.__Ownable_init_unchained();
+
+        if (_taikoToken == address(0)) revert INVALID_PARAM();
+        taikoToken = _taikoToken;
+
+        if (_sharedVault == address(0)) revert INVALID_PARAM();
+        sharedVault = _sharedVault;
```

Comment:
Maybe a setter for `sharedVault` ? just in case vault is compromised somehow ? 🤔